File: app/src/main/java/com/example/myapplication/network/NetworkService.java
package com.example.myapplication.network;

import android.content.Context;
import android.util.Log;

import com.example.myapplication.auth.AuthManager;
import com.example.myapplication.dto.LoginRequest;
import com.example.myapplication.dto.ApiResponse;
import com.example.myapplication.dto.LoginResponse;
import com.example.myapplication.dto.UserDTO;
import com.example.myapplication.dto.PostDTO;
import com.example.myapplication.dto.MessageDTO;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class NetworkService {
    private static final String TAG = "NetworkService";
    private static NetworkService instance;
    private ApiService apiService;
    private AuthManager authManager;
    
    private NetworkService(Context context) {
        apiService = NetworkConfig.getApiService();
        authManager = AuthManager.getInstance(context);
    }
    
    public static synchronized NetworkService getInstance(Context context) {
        if (instance == null) {
            instance = new NetworkService(context.getApplicationContext());
        }
        return instance;
    }
    
    
    // 发送验证码
    public void sendVerificationCode(String phone, NetworkCallback<String> callback) {
        apiService.sendVerificationCode(phone).enqueue(new Callback<ApiResponse<String>>() {
            @Override
            public void onResponse(Call<ApiResponse<String>> call, Response<ApiResponse<String>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    ApiResponse<String> apiResponse = response.body();
                    if (apiResponse.isSuccess()) {
                        callback.onSuccess(apiResponse.getData());
                    } else {
                        callback.onError(apiResponse.getMessage());
                    }
                } else {
                    callback.onError("发送验证码失败: " + response.message());
                }
            }
            
            @Override
            public void onFailure(Call<ApiResponse<String>> call, Throwable t) {
                Log.e(TAG, "发送验证码请求失败", t);
                callback.onError("网络错误: " + t.getMessage());
            }
        });
    }
    
    // 验证码登录/注册
    public void loginWithVerificationCode(String phone, String code, NetworkCallback<LoginResponse> callback) {
        apiService.loginWithVerificationCode(phone, code).enqueue(new Callback<ApiResponse<LoginResponse>>() {
            @Override
            public void onResponse(Call<ApiResponse<LoginResponse>> call, Response<ApiResponse<LoginResponse>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    ApiResponse<LoginResponse> apiResponse = response.body();
                    if (apiResponse.isSuccess()) {
                        LoginResponse loginResponse = apiResponse.getData();
                        // 保存认证信息
                        authManager.saveToken(loginResponse.getToken());
                        authManager.saveUserId(loginResponse.getUser().getId());
                        authManager.saveUsername(loginResponse.getUser().getUsername());
                        callback.onSuccess(loginResponse);
                    } else {
                        callback.onError(apiResponse.getMessage());
                    }
                } else {
                    callback.onError("登录失败: " + response.message());
                }
            }
            
            @Override
            public void onFailure(Call<ApiResponse<LoginResponse>> call, Throwable t) {
                Log.e(TAG, "验证码登录请求失败", t);
                callback.onError("网络错误: " + t.getMessage());
            }
        });
    }
    
    // 用户登录
    public void login(LoginRequest request, NetworkCallback<LoginResponse> callback) {
        apiService.login(request).enqueue(new Callback<ApiResponse<LoginResponse>>() {
            @Override
            public void onResponse(Call<ApiResponse<LoginResponse>> call, Response<ApiResponse<LoginResponse>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    ApiResponse<LoginResponse> apiResponse = response.body();
                    if (apiResponse.isSuccess()) {
                        LoginResponse loginResponse = apiResponse.getData();
                        // 保存认证信息
                        authManager.saveToken(loginResponse.getToken());
                        authManager.saveUserId(loginResponse.getUser().getId());
                        authManager.saveUsername(loginResponse.getUser().getUsername());
                        callback.onSuccess(loginResponse);
                    } else {
                        callback.onError(apiResponse.getMessage());
                    }
                } else {
                    callback.onError("登录失败: " + response.message());
                }
            }
            
            @Override
            public void onFailure(Call<ApiResponse<LoginResponse>> call, Throwable t) {
                Log.e(TAG, "登录请求失败", t);
                callback.onError("网络错误: " + t.getMessage());
            }
        });
    }
    
    // 获取用户信息
    public void getProfile(NetworkCallback<UserDTO> callback) {
        String authHeader = authManager.getAuthHeader();
        if (authHeader == null) {
            callback.onError("未登录");
            return;
        }
        
        apiService.getProfile(authHeader).enqueue(new Callback<ApiResponse<UserDTO>>() {
            @Override
            public void onResponse(Call<ApiResponse<UserDTO>> call, Response<ApiResponse<UserDTO>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    ApiResponse<UserDTO> apiResponse = response.body();
                    if (apiResponse.isSuccess()) {
                        callback.onSuccess(apiResponse.getData());
                    } else {
                        callback.onError(apiResponse.getMessage());
                    }
                } else {
                    callback.onError("获取用户信息失败: " + response.message());
                }
            }
            
            @Override
            public void onFailure(Call<ApiResponse<UserDTO>> call, Throwable t) {
                Log.e(TAG, "获取用户信息请求失败", t);
                callback.onError("网络错误: " + t.getMessage());
            }
        });
    }
    
    // 搜索用户
    public void searchUsers(String keyword, String gender, String location, 
                          Integer minAge, Integer maxAge, int page, int size, 
                          NetworkCallback<List<UserDTO>> callback) {
        apiService.searchUsers(keyword, gender, location, minAge, maxAge, page, size)
                .enqueue(new Callback<ApiResponse<List<UserDTO>>>() {
                    @Override
                    public void onResponse(Call<ApiResponse<List<UserDTO>>> call, Response<ApiResponse<List<UserDTO>>> response) {
                        if (response.isSuccessful() && response.body() != null) {
                            ApiResponse<List<UserDTO>> apiResponse = response.body();
                            if (apiResponse.isSuccess()) {
                                callback.onSuccess(apiResponse.getData());
                            } else {
                                callback.onError(apiResponse.getMessage());
                            }
                        } else {
                            callback.onError("搜索用户失败: " + response.message());
                        }
                    }
                    
                    @Override
                    public void onFailure(Call<ApiResponse<List<UserDTO>>> call, Throwable t) {
                        Log.e(TAG, "搜索用户请求失败", t);
                        callback.onError("网络错误: " + t.getMessage());
                    }
                });
    }
    
    // 获取动态列表
    public void getPosts(int page, int size, NetworkCallback<List<PostDTO>> callback) {
        apiService.getPosts(page, size).enqueue(new Callback<ApiResponse<List<PostDTO>>>() {
            @Override
            public void onResponse(Call<ApiResponse<List<PostDTO>>> call, Response<ApiResponse<List<PostDTO>>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    ApiResponse<List<PostDTO>> apiResponse = response.body();
                    if (apiResponse.isSuccess()) {
                        callback.onSuccess(apiResponse.getData());
                    } else {
                        callback.onError(apiResponse.getMessage());
                    }
                } else {
                    callback.onError("获取动态列表失败: " + response.message());
                }
            }
            
            @Override
            public void onFailure(Call<ApiResponse<List<PostDTO>>> call, Throwable t) {
                Log.e(TAG, "获取动态列表请求失败", t);
                callback.onError("网络错误: " + t.getMessage());
            }
        });
    }
    
    // 点赞动态
    public void likePost(Long postId, NetworkCallback<PostDTO> callback) {
        String authHeader = authManager.getAuthHeader();
        if (authHeader == null) {
            callback.onError("未登录");
            return;
        }
        
        apiService.likePost(authHeader, postId).enqueue(new Callback<ApiResponse<PostDTO>>() {
            @Override
            public void onResponse(Call<ApiResponse<PostDTO>> call, Response<ApiResponse<PostDTO>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    ApiResponse<PostDTO> apiResponse = response.body();
                    if (apiResponse.isSuccess()) {
                        callback.onSuccess(apiResponse.getData());
                    } else {
                        callback.onError(apiResponse.getMessage());
                    }
                } else {
                    callback.onError("点赞失败: " + response.message());
                }
            }
            
            @Override
            public void onFailure(Call<ApiResponse<PostDTO>> call, Throwable t) {
                Log.e(TAG, "点赞请求失败", t);
                callback.onError("网络错误: " + t.getMessage());
            }
        });
    }
    
    // 网络回调接口
    public interface NetworkCallback<T> {
        void onSuccess(T data);
        void onError(String error);
    }
}

File: app/src/main/java/com/example/myapplication/network/NetworkService.kt
package com.example.myapplication.network

import com.example.myapplication.auth.AuthManager
import com.example.myapplication.dto.ApiResponse
import com.example.myapplication.dto.WalletDTO
import com.example.myapplication.dto.UserPhotoDTO
import com.example.myapplication.dto.UploadPhotoResponse
import com.example.myapplication.dto.AlipayOrderResponse
import com.example.myapplication.dto.CreateOrderRequest
import com.example.myapplication.dto.PaymentOrderDTO
import com.example.myapplication.dto.UserSettingsDTO
import retrofit2.Response
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MultipartBody
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.asRequestBody
import java.io.File

object NetworkService {
    
    suspend fun getWalletBalance(token: String): Response<ApiResponse<WalletDTO>> {
        return withContext(Dispatchers.IO) {
            try {
                val call = NetworkConfig.getApiService().getWalletBalance("Bearer $token")
                call.execute()
            } catch (e: Exception) {
                throw e
            }
        }
    }

    // 用户设置相关方法

    /**
     * 获取用户设置
     */
    suspend fun getUserSettings(token: String): Response<ApiResponse<UserSettingsDTO>> {
        return withContext(Dispatchers.IO) {
            try {
                val call = NetworkConfig.getApiService().getUserSettings("Bearer $token")
                call.execute()
            } catch (e: Exception) {
                throw e
            }
        }
    }

    /**
     * 更新用户设置
     */
    suspend fun updateUserSettings(token: String, settings: UserSettingsDTO): Response<ApiResponse<UserSettingsDTO>> {
        return withContext(Dispatchers.IO) {
            try {
                val call = NetworkConfig.getApiService().updateUserSettings("Bearer $token", settings)
                call.execute()
            } catch (e: Exception) {
                throw e
            }
        }
    }
    
    // 支付相关方法
    
    /**
     * 创建支付宝订单
     */
    suspend fun createAlipayOrder(token: String, request: CreateOrderRequest): Response<ApiResponse<AlipayOrderResponse>> {
        return withContext(Dispatchers.IO) {
            try {
                val call = NetworkConfig.getApiService().createAlipayOrder("Bearer $token", request)
                call.execute()
            } catch (e: Exception) {
                throw e
            }
        }
    }
    
    /**
     * 获取订单列表
     */
    suspend fun getOrderList(token: String, status: String? = null): Response<ApiResponse<List<PaymentOrderDTO>>> {
        return withContext(Dispatchers.IO) {
            try {
                val call = NetworkConfig.getApiService().getOrderList("Bearer $token", status)
                call.execute()
            } catch (e: Exception) {
                throw e
            }
        }
    }
    
    /**
     * 获取订单详情
     */
    suspend fun getOrderDetail(token: String, orderId: String): Response<ApiResponse<PaymentOrderDTO>> {
        return withContext(Dispatchers.IO) {
            try {
                val call = NetworkConfig.getApiService().getOrderDetail("Bearer $token", orderId)
                call.execute()
            } catch (e: Exception) {
                throw e
            }
        }
    }
    
    /**
     * 取消订单
     */
    suspend fun cancelOrder(token: String, orderId: String): Response<ApiResponse<String>> {
        return withContext(Dispatchers.IO) {
            try {
                val call = NetworkConfig.getApiService().cancelOrder("Bearer $token", orderId)
                call.execute()
            } catch (e: Exception) {
                throw e
            }
        }
    }
    
    /**
     * 获取支付统计
     */
    suspend fun getPaymentStatistics(token: String): Response<ApiResponse<Map<String, Any>>> {
        return withContext(Dispatchers.IO) {
            try {
                val call = NetworkConfig.getApiService().getPaymentStatistics("Bearer $token")
                call.execute()
            } catch (e: Exception) {
                throw e
            }
        }
    }
    
    // 相册相关方法
    
    /**
     * 获取用户相册照片列表
     */
    suspend fun getUserPhotos(userId: Long, context: android.content.Context): Response<ApiResponse<List<UserPhotoDTO>>> {
        return withContext(Dispatchers.IO) {
            try {
                val authManager = AuthManager.getInstance(context)
                val authHeader = "Bearer ${authManager.getToken()}"
                val call = NetworkConfig.getApiService().getUserPhotos(userId, authHeader)
                call.execute()
            } catch (e: Exception) {
                throw e
            }
        }
    }
    
    /**
     * 上传照片到相册
     */
    suspend fun uploadPhoto(userId: Long, photoFile: File, context: android.content.Context, isAvatar: Boolean = false): Response<ApiResponse<UploadPhotoResponse>> {
        return withContext(Dispatchers.IO) {
            try {
                val authManager = AuthManager.getInstance(context)
                val authHeader = "Bearer ${authManager.getToken()}"
                val requestFile = photoFile.asRequestBody("image/*".toMediaType())
                val photoPart = MultipartBody.Part.createFormData("photo", photoFile.name, requestFile)
                val call = NetworkConfig.getApiService().uploadPhoto(userId, photoPart, isAvatar, authHeader)
                call.execute()
            } catch (e: Exception) {
                throw e
            }
        }
    }
    
    /**
     * 删除照片
     */
    suspend fun deletePhoto(userId: Long, photoId: Long, context: android.content.Context): Response<ApiResponse<String>> {
        return withContext(Dispatchers.IO) {
            try {
                val authManager = AuthManager.getInstance(context)
                val authHeader = "Bearer ${authManager.getToken()}"
                val call = NetworkConfig.getApiService().deletePhoto(userId, photoId, authHeader)
                call.execute()
            } catch (e: Exception) {
                throw e
            }
        }
    }
    
    /**
     * 设置照片为头像
     */
    suspend fun setAsAvatar(userId: Long, photoId: Long, context: android.content.Context): Response<ApiResponse<String>> {
        return withContext(Dispatchers.IO) {
            try {
                val authManager = AuthManager.getInstance(context)
                val authHeader = "Bearer ${authManager.getToken()}"
                val call = NetworkConfig.getApiService().setAsAvatar(userId, photoId, authHeader)
                call.execute()
            } catch (e: Exception) {
                throw e
            }
        }
    }
    
    // 身份证二要素核验相关方法
    
    /**
     * 提交身份证二要素核验
     */
    suspend fun submitIdCardVerification(certName: String, certNo: String, context: android.content.Context): Map<String, Any> {
        return withContext(Dispatchers.IO) {
            try {
                val authManager = AuthManager.getInstance(context)
                val authHeader = "Bearer ${authManager.getToken()}"
                val request = mapOf(
                    "certName" to certName,
                    "certNo" to certNo
                )
                val call = NetworkConfig.getApiService().submitIdCardVerification(authHeader, request)
                val response = call.execute()
                
                if (response.isSuccessful) {
                    val apiResponse = response.body()
                    if (apiResponse?.isSuccess() == true) {
                        mapOf(
                            "success" to true,
                            "message" to apiResponse.message,
                            "verificationResult" to apiResponse.data
                        )
                    } else {
                        mapOf(
                            "success" to false,
                            "message" to (apiResponse?.message ?: "认证失败")
                        )
                    }
                } else {
                    mapOf(
                        "success" to false,
                        "message" to "网络请求失败: ${response.code()}"
                    )
                }
            } catch (e: Exception) {
                mapOf(
                    "success" to false,
                    "message" to "网络错误: ${e.message}"
                )
            }
        }
    }
    
    /**
     * 获取实名认证状态
     */
    suspend fun getVerificationStatus(context: android.content.Context): Map<String, Any> {
        return withContext(Dispatchers.IO) {
            try {
                val authManager = AuthManager.getInstance(context)
                val authHeader = "Bearer ${authManager.getToken()}"
                val call = NetworkConfig.getApiService().getVerificationStatus(authHeader)
                val response = call.execute()
                
                if (response.isSuccessful) {
                    val apiResponse = response.body()
                    if (apiResponse?.isSuccess() == true) {
                        mapOf(
                            "success" to true,
                            "data" to apiResponse.data
                        )
                    } else {
                        mapOf(
                            "success" to false,
                            "message" to (apiResponse?.message ?: "查询失败")
                        )
                    }
                } else {
                    mapOf(
                        "success" to false,
                        "message" to "网络请求失败: ${response.code()}"
                    )
                }
            } catch (e: Exception) {
                mapOf(
                    "success" to false,
                    "message" to "网络错误: ${e.message}"
                )
            }
        }
    }
    
    /**
     * 查询认证结果
     */
    suspend fun getVerificationResult(context: android.content.Context): Map<String, Any> {
        return withContext(Dispatchers.IO) {
            try {
                val authManager = AuthManager.getInstance(context)
                val authHeader = "Bearer ${authManager.getToken()}"
                val call = NetworkConfig.getApiService().getVerificationResult(authHeader)
                val response = call.execute()
                
                if (response.isSuccessful) {
                    val apiResponse = response.body()
                    if (apiResponse?.isSuccess() == true) {
                        mapOf(
                            "success" to true,
                            "data" to apiResponse.data
                        )
                    } else {
                        mapOf(
                            "success" to false,
                            "message" to (apiResponse?.message ?: "查询失败")
                        )
                    }
                } else {
                    mapOf(
                        "success" to false,
                        "message" to "网络请求失败: ${response.code()}"
                    )
                }
            } catch (e: Exception) {
                mapOf(
                    "success" to false,
                    "message" to "网络错误: ${e.message}"
                )
            }
        }
    }
}
